Automattic/jetpack
# Sync: Migrate 'jetpack_sync_before_send*' actions for Sync queue to 'jetpack_sync_before_enqueue' instead

## Proposed changes:
- Migrates `jetpack_sync_before_send_*` actions for the Sync queue to `jetpack_sync_before_enqueue_*` instead.
- ~Updates the Dedicated Sync init hook priority to `0` (was `90`)~ Will do in a follow up PR 

The idea behind this change is to make the Dedicated Sync requests as fast as possible. When a Dedicated Sync request is triggered (Sync queue is unblocked and non empty, aka we need to send those actions to WPCOM), we fetch the Sync queue and send the corresponding actions to WPCOM. Ideally we want this done as fast as possible, thus independent from any other plugin `init` hooks so it can execute and exit early.
However, the current logic behind `jetpack_sync_before_send_*` actions prevented us from doing so. An example is the Woo regression we introduced when we attempted to only update Dedicated Sync's `init` hook priority in the past. 
By only updating the `init` hook priority we ended up with Cache site posts (orders) with `jps_non-reg` statuses and missing fields. See p1HpG7-nxq-p2
This happened because we were essentially checking/updating those queued posts on `jetpack_sync_before_send_jetpack_sync_action` but the corresponding post type (`orders`) was not registered yet when the Dedicated Sync hook priority was updated to `0`.
Therefore this PR moves the corresponding actions that would modify content before sending to before enqueueing instead.

**Pros:**
- More accurate sending of data, since the synced items might have changed between enqueuing and sending
- ~Faster sending of Sync actions to WPCOMFaster sending of Sync actions to WPCOM~ Will do in a follow up PR (related to Dedicated Sync `init` hook priority)
- ~Fixes [Cart contents cleared when logging in at checkout w/ WCPay and JP active 
](https://github.com/Automattic/woocommerce-payments/issues/5491)~ Will do in a follow up PR (related to Dedicated Sync `init` hook priority)

**Cons:**
- Since we are expanding some actions before adding them to the queue, their corresponding size in the DB is going to grow

### Other information:


## Jetpack product discussion


p9dueE-7zI-p2

## Does this pull request change what data or activity we track or use?


No

## Testing instructions:
- Ensure Sync actions are properly synced for the affected modules: Posts, Comments, Users, Updates, Plugins
- Ensure checksums are still valid for the affected modules: Posts, Comments, Users, Updates, Plugins
- Test with WooCommerce and ensure Orders are successfully synced

_A few specific cases TBA / See unit tests_
